
extractor.bowtie2_align <- function(bowtie2_command, threads, currentSample, extractedFastqDirectory){
  currentSample$samPath <- file.path(extractedFastqDirectory,paste0(currentSample$name,'.sam'))
  formattedDate <- format(Sys.Date(), "%m%d%Y")
  optionsCommand <- c('-x Resources/extractor_resources/reference/output',
                      '-5 3','-3 7','-L 20','-i S,1,0.5','--score-min L,0,-0.187',
                      '-I 75','-X 1000', 
                      paste0('-p ',threads),
                      paste0('-1 ',currentSample$rawfastq1path),
                      paste0('-2 ',currentSample$rawfastq2path),
                      paste0('-S ',currentSample$samPath),
                      paste0('--al-conc-gz ',
                             file.path(extractedFastqDirectory,paste0(currentSample$name,'%_001-HLA-',formattedDate,'.fastq.gz'))
                      ),
                      '--un delete.me'
  )
  cat('\n\n',bowtie2_command, optionsCommand)
  output.sampleAlign <- system2(bowtie2_command, optionsCommand, stdout=T, stderr=T)
  
  if(!is.null(attributes(output.sampleAlign))){
    cat('\nBowtie2 alignment failed, retrying...')
    output.sampleAlign <- system2(bowtie2_command, optionsCommand, stdout=T, stderr=T)
  }
  
  check.system2_output(output.sampleAlign, 'Bowtie2 HLA extraction alignment failed.')
  
  cat('\n',paste0(output.sampleAlign, collapse='\n'))
  
  
  currentSample[['HLAfastq1path']] <- normalizePath(file.path(extractedFastqDirectory,paste0(currentSample$name,'1_001-HLA-',formattedDate,'.fastq.gz')))
  currentSample[['HLAfastq2path']] <- normalizePath(file.path(extractedFastqDirectory,paste0(currentSample$name,'2_001-HLA-',formattedDate,'.fastq.gz')))
  
  cat('\n\nSuccessfully extracted HLA reads for',currentSample$name)
  
  cat('\nCleaning up alignment files.')
  file.remove('delete.me')
  file.remove(currentSample$samPath)
  
  return(currentSample)
}

extractor.run <- function(sampleList, threads, extractedFastqDirectory, forceRun=F){
  
  # Create the directory if it does not yet exist
  if(!file.exists(extractedFastqDirectory)){
    dir.create(extractedFastqDirectory,recursive=T)
    cat('\n',extractedFastqDirectory,'created.\n')
  }else{
    cat('\n',extractedFastqDirectory,'found.\n')
  }
  
  for(currentSample in sampleList){
    cat('\n\nProcessing',currentSample$name,'----------')
    
    ## forceRun=T forces the alignment to run, even if there are previous results
    if(!forceRun){
      previousResultVect <- grep(currentSample$name, list.files(extractedFastqDirectory,full.names=T),value=T)
      
      ## If there are two files found and they both have _HLA_ in them, 
      #  add the file names to the sample object and skip alignment
      if(length(previousResultVect) == 2 & length(grep('_HLA_',previousResultVect,fixed=T)) == 2){
        cat('\nExtracted files found, skipping alignment.')
        currentSample[['HLAfastq1path']] <- grep('HLA_1.fastq.gz',previousResultVect,value=T)
        currentSample[['HLAfastq2path']] <- grep('HLA_2.fastq.gz',previousResultVect,value=T)
        
        next
      }
    }
    
    currentSample <- extractor.bowtie2_align(bowtie2, threads, currentSample, extractedFastqDirectory)
  }
  
  cat("\n\n----- PING2_extractor is complete. Extracted reads are deposited in",extractedFastqDirectory,'-----\n')
  
  return(sampleList)
}
